packages/data/src/user-preferences/use-user-preferences.js

```diff
@@ -0,0 +1,149 @@
+	} );
+
+	return userData;
+};
+
+/**
+ * Custom react hook for retrieving thecurrent user's WooCommerce preferences.
+ *
+ * This is a wrapper around @wordpress/core-data's getCurrentUser() and saveUser().
+ */
+export const useUserPreferences = () => {
+	// Get our dispatch methods now - this can't happen inside the callback below.
+	const { receiveCurrentUser, saveUser } = useDispatch( STORE_NAME );
+
+	const { isRequesting, userPreferences, updateUserPreferences } = useSelect(
+		( select ) => {
+			const {
+				getCurrentUser,
+				getLastEntitySaveError,
+				hasStartedResolution,
+				hasFinishedResolution,
+			} = select( STORE_NAME );
+
+			// Use getCurrentUser() to get WooCommerce meta values.
+			const user = getCurrentUser();
+			const userData = getWooCommerceMeta( user );
+
+			// Create wrapper for updating user's `woocommerce_meta`.
+			const updateUserPrefs = async ( userPrefs ) => {
+				// @todo Handle unresolved getCurrentUser() here.
+
+				// Whitelist our meta fields.
+				const userDataFields = [
+					'categories_report_columns',
+					'coupons_report_columns',
+					'customers_report_columns',
+					'orders_report_columns',
+					'products_report_columns',
+					'revenue_report_columns',
+					'taxes_report_columns',
+					'variations_report_columns',
+					'dashboard_sections',
+					'dashboard_chart_type',
+					'dashboard_chart_interval',
+					'dashboard_leaderboard_rows',
+					'activity_panel_inbox_last_read',
+					'homepage_stats',
+				];
+
+				// Prep valid fields for update.
+				const metaData = mapValues(
+					pick( userPrefs, userDataFields ),
+					JSON.stringify
+				);
+
+				if ( Object.keys( metaData ).length === 0 ) {
+					return {
+						error: new Error(
+							'No valid woocommerce_meta keys were provided for update.'
+						),
+						updatedUser: undefined,
+					};
+				}
+
+				// Use saveUser() to update WooCommerce meta values.
+				const updatedUser = await saveUser( {
+					id: user.id,
+					woocommerce_meta: metaData,
+				} );
+
+				if ( undefined === updatedUser ) {
+					// Return the encountered error to the caller.
+					const error = getLastEntitySaveError(
+						'root',
+						'user',
+						user.id
+					);
+
+					return {
+						error,
+						updatedUser,
+					};
+				}
+
+				// Propagate the updated User object to the store.
+				receiveCurrentUser( updatedUser );
+
+				// Decode the WooCommerce meta after save.
+				const updatedUserResponse = {
+					...updatedUser,
+					woocommerce_meta: getWooCommerceMeta( updatedUser ),
+				};
+
+				// @todo: Do we need to start/finish resolution here?
+				// E.g.:
+				// startResolution( 'getCurrentUser', [] );
+				// receiveCurrentUser( userRef.current );
+				// finishResolution( 'getCurrentUser', [] );
+
+				return {
+					updatedUser: updatedUserResponse,
+				};
+			};
+
+			return {
+				isRequesting:
+					hasStartedResolution( 'getCurrentUser' ) &&
+					! hasFinishedResolution( 'getCurrentUser' ),
```

Comment:
Going with the comment below, you should also add empty array here as second arg. I like the abstraction here though.

Reply:
The empty array argument is unnecessary. See: https://github.com/woocommerce/woocommerce-admin/pull/4505#discussion_r436019544